repo/grpc_repository_client.go

```diff
@@ -477,56 +625,107 @@ func OpenGRPCAPIRepository(ctx context.Context, si *APIServerInfo, cliOpts Clien
 		return nil, errors.Wrap(err, "dial error")
 	}
 
-	rep, err := NewGRPCAPIRepositoryForConnection(ctx, conn, new(int32), cliOpts, "")
+	rep, err := newGRPCAPIRepositoryForConnection(ctx, conn, new(int32), cliOpts, "", true)
 	if err != nil {
 		return nil, err
 	}
 
 	return rep, nil
 }
 
-// NewGRPCAPIRepositoryForConnection opens GRPC-based repository connection.
-func NewGRPCAPIRepositoryForConnection(ctx context.Context, conn *grpc.ClientConn, connRefCount *int32, cliOpts ClientOptions, purpose string) (RepositoryWriter, error) {
-	cli := apipb.NewKopiaRepositoryClient(conn)
+func (r *grpcRepositoryClient) getOrEstablishInnerSession(ctx context.Context) (*grpcInnerSession, error) {
+	r.innerSessionMutex.Lock()
+	defer r.innerSessionMutex.Unlock()
 
-	sess, err := cli.Session(ctx)
-	if err != nil {
-		return nil, errors.Wrap(err, "error starting session")
-	}
+	if r.innerSession == nil {
+		cli := apipb.NewKopiaRepositoryClient(r.conn)
 
-	rr := &grpcRepositoryClient{
-		connRefCount:   connRefCount,
-		conn:           conn,
-		cliOpts:        cliOpts,
-		sess:           sess,
-		activeRequests: make(map[int64]chan *apipb.SessionResponse),
-		nextRequestID:  1,
-	}
+		log(ctx).Debugf("establishing new GRPC streaming session")
 
-	go rr.readLoop()
+		retryPolicy := retry.Always
+		if !r.hadInnerSession && r.transparentRetries {
+			// the first time the read-only session is established, don't do retries
+			// to avoid spinning in place while the server is not connectable.
+			retryPolicy = retry.Never
+		}
 
-	p, err := rr.initializeSession(purpose, cliOpts.ReadOnly)
-	if err != nil {
-		return nil, errors.Wrap(err, "unable to initialize session")
-	}
+		v, err := retry.WithExponentialBackoff(ctx, "establishing session", func() (interface{}, error) {
+			sess, err := cli.Session(ctxutil.Detach(ctx))
+			if err != nil {
+				return nil, errors.Wrap(err, "Session()")
+			}
+
+			newSess := &grpcInnerSession{
+				cli:            sess,
+				activeRequests: make(map[int64]chan *apipb.SessionResponse),
+				nextRequestID:  1,
+			}
+
+			go newSess.readLoop(ctx)
+
+			newSess.repoParams, err = newSess.initializeSession(ctx, r.purpose, r.isReadOnly)
+			if err != nil {
+				return nil, errors.Wrap(err, "unable to initialize session")
+			}
+
+			return newSess, nil
+		}, retryPolicy)
+		if err != nil {
+			return nil, errors.Wrap(err, "error establishing session")
+		}
 
-	hf, err := hashing.CreateHashFunc(p)
-	if err != nil {
-		return nil, errors.Wrap(err, "unable to create hash function")
+		r.innerSession = v.(*grpcInnerSession)
 	}
 
-	rr.h = hf
+	return r.innerSession, nil
+}
+
+func (r *grpcRepositoryClient) killInnerSession() {
+	r.innerSessionMutex.Lock()
+	defer r.innerSessionMutex.Unlock()
 
-	rr.objectFormat = object.Format{
-		Splitter: p.Splitter,
+	if r.innerSession != nil {
+		r.innerSession.cli.CloseSend() //nolint:errcheck
+		r.innerSession = nil
```

Comment:
Do errors need to be sent to the channels for active requests and then closed? Would active requests hang otherwise?

Reply:
the read loop will terminate because the server will also disconnect in this case and will stuff errors into reader channels.